internal/pkg/service/buffer/storage/local/writer/writechain/chain.go

```diff
@@ -187,13 +177,8 @@ func (c *Chain) PrependWriterOrErr(factory func(Writer) (io.Writer, error)) (ok
 	if !same {
 		c.writers = append([]io.Writer{newWriter}, c.writers...)
 
-		// Check if the new writer implements WriteString method
-		if v, ok := newWriter.(Writer); ok {
-			c.beginning = v
-		} else {
-			// Add WriteString method, if the new writer doesn't implement it.
-			c.beginning = &stringWriter{Writer: newWriter}
-		}
+		// Protect asynchronous Flush with a lock
+		safe := newSafeWriter(newWriter)
```

Comment:
**4. problém** bol pri writers, ktoré majú metódu `Write` a `Flush`. Chýbal zámok.
- Je zabezpečené, že `Write` sa na tejto úrovni nevolá paralelne, je tam zámok hned na začiatku.
- Ale `Flush` (časť `Sync`) sa volá asynchrónne pri splnení podmienok.

Writers v Go by default vačšinou nemajú locks, ak nie je uvedené inak. Kedže často nepotrebuješ paralený zapís/flush, my to ale potrebujeme.

Preto každý writer obalíme pomocou `newSafeWriter` -> `safeWriter(Write,WriteString,Flush)`.
- Ak povodny writer nemal metodu `Flush`, tak ho nemusíme pridat medzi `flushers`.
- Zámok nad `Writer` ale potrebujeme vždy, lebo `Flush` niektorého predchádzajúceho witer, môže sposobiť paralelné zavolanie `Write` metódy niekde neskôr.